### ISSUE TITLE
Placeholder emails not going out

### ISSUE THREAD

**Issue by ianlayton** [MidwestFurryFandom/mff-rams-plugin#40]
Emails that should be going out when creating a placeholder registration (for rollovers and staff) don't seem to be going out. 

http://rams.test.furfest.org/rams/registration/form?id=cbcec10c-d132-4699-aeeb-c4a0f868f6cb

Email is actually showing up in the email admin section. This was done after it was marked as approved:

http://rams.test.furfest.org/rams/emails/pending_examples?ident=regular_badge_confirmation

**Comment by kitsuta** [comment]
It does seem as though this email is just not being sent. Even on my local copy, where it recognized that there were emails waiting to be sent, approving them didn't seem to _do_ anything. @RobRuana can we touch base on why these emails aren't going out, even though the system knows they ought to be? Here's what it looks like for me:
![image](https://user-images.githubusercontent.com/7198215/40463338-bb77b7bc-5ee2-11e8-9dde-fe8de3181554.png)
![image](https://user-images.githubusercontent.com/7198215/40463347-c7319758-5ee2-11e8-85c8-cffe67301f6e.png)

There don't seem to be any relevant logs for the celery worker, celery beat, or web containers.

**Comment by RobRuana** [comment]
Hey, I just took a look at the pages linked above (on http://rams.test.furfest.org), and it _looks_ like the emails were sent. Did this get fixed? Were the celery processes not running?

**Comment by kitsuta** [comment]
No, the emails you see in the user's history are the _payment confirmation_ emails for other badges. The email that's not being sent is the placeholder badge confirmation email, titled "Midwest FurFest Badge Confirmation". In the staging server, the email process is not even seeing that there are emails _TO_ send, despite there definitely being placeholder comped Attendee badges without ribbons.

**Comment by ianlayton** [comment]
This issue can't be put off until iteration 2 because we need to be able to send out these place holder emails as soon as we launch because of the hotel lottery.

**Comment by jmdawson** [comment]
Ok, since this affects >100 users between now and before-iteration2, let's see if we can deliver this either before iteration 1 launches or shortly thereafter.

**Comment by kitsuta** [comment]
@RobRuana

**Comment by kitsuta** [comment]
Good news is, we got what seems to be a workaround: Setting the email to not need approval (`needs_approval=False,`) causes it to send correctly. The tricky part is that just changing the code isn't sufficient -- I had to completely wipe my docker containers and re-up before the email would change to being automatic.

**Comment by RobRuana** [comment]
@kitsuta I am digging into this right now. Feel free to ping me anytime you run into issues like this. If it's affecting mff, it's also probably affecting MAGFest!

**Comment by RobRuana** [comment]
Quick side note: there is a `sep send_automated_emails` command that is included for dev purposes. It does the same thing as the `send_automated_emails` celery task, but you can kick it off from the command line. When I'm troubleshooting email issues, I usually shut down the celery services and run `sep send_automated_emails` from the command line. If that's working, then I can narrow it down to a problem with the service.

Okay, this _appears_ to be working as expected on my local development machine, as well as our production servers. That leads me to suspect it has something to do with your docker configuration interacting poorly with the system's startup routine. I haven't done any work to make the startup routine bulletproof, since we kinda expect all of our services (postgres, rabbit, etc..) to be running when the system starts up.

I'd start by running `sep send_automated_emails` and verifying that works as expected. Then I'd inspect the celery logs when you restart your containers (on our systems they're located at `/var/log/uber-celery-worker.log` and `/var/log/uber-celery-beat.log`). The startup for the worker should look something like this:
```
[2018-05-25 10:21:41,382: INFO/MainProcess] Received task: uber.models.email.reconcile_fixtures[508e30fc-343c-40dd-a2dd-36e154dc622b]  
[2018-05-25 10:21:42,028: INFO/ForkPoolWorker-2] Task uber.models.email.reconcile_fixtures[508e30fc-343c-40dd-a2dd-36e154dc622b] succeeded in 0.5407586390028882s: None
[2018-05-25 10:22:09,899: INFO/MainProcess] Received task: uber.tasks.email.send_automated_emails[28a26bd4-f495-473b-bf06-4bc65bb946ad]  
[2018-05-25 10:22:10,022: INFO/ForkPoolWorker-2] Task uber.tasks.email.send_automated_emails[28a26bd4-f495-473b-bf06-4bc65bb946ad] succeeded in 0.12200422499881824s: {'badge_confirmation_reminder': 1, 'regular_badge_confirmation': 2}
```
If you see any errors in the logs, that's a huge red flag. If you **DON'T** see `uber.models.email.reconcile_fixtures` that's **ALSO** a huge red flag.

**NOTE**: the output of `uber.tasks.email.send_automated_emails` is supposed to indicate the number of _unapproved_ emails for each category _that would have been sent if they were approved_. In this case, I had not yet approved `regular_badge_confirmation`.

After startup, I then approved `regular_badge_confirmation` and waited for `uber.tasks.email.send_automated_emails` to run again. 
```
[2018-05-25 10:27:09,952: INFO/MainProcess] Received task: uber.tasks.email.send_automated_emails[61e4a966-5b0a-4438-afab-1416fb06a315]  
[2018-05-25 10:27:10,083: ERROR/ForkPoolWorker-2] Email sending turned off, so unable to send {'original_cc': [], 'to': [], 'subject': 'MAGStock Badge Confirmation', 'original_bcc': [], 'session': None, 'sender': 'MAGStock Registration <admin@magstock.org>', 'ident': 'regular_badge_confirmation', 'model': {'id': '4c5ec46c-667a-42dd-986a-ba03522823a3', '_model': 'Attendee'}, 'format': 'text', 'original_to': ['magfest@example.com'], 'automated_email': {'id': 'ca4b544f-32f3-489f-87ef-cb1e917718e4', '_model': 'AutomatedEmail'}, 'cc': [], 'bcc': [], 'body': "Test,\n\nYou've been added to the MAGStock registration database, but we don't have all of your personal information and you have not yet paid your outstanding balance of $60.  To ensure that you can pick up your badge with no hassles at our registration desk, please fill out the rest of your info and pay at https://vagrant-ubuntu-trusty-64.lan:4443/uber/preregistration/confirm?id=4c5ec46c-667a-42dd-986a-ba03522823a3 and then simply bring a photo ID to MAGStock.\n\nPlease let us know if you have any questions.\n\nMAGStock Registration Department"}
[2018-05-25 10:27:10,103: ERROR/ForkPoolWorker-2] Email sending turned off, so unable to send {'original_cc': [], 'to': [], 'subject': 'MAGStock Badge Confirmation', 'original_bcc': [], 'session': None, 'sender': 'MAGStock Registration <admin@magstock.org>', 'ident': 'regular_badge_confirmation', 'model': {'id': '157712d4-aa9a-49f3-904e-a59be6933d84', '_model': 'Attendee'}, 'format': 'text', 'original_to': ['testplaceholderemail@example.com'], 'automated_email': {'id': 'ca4b544f-32f3-489f-87ef-cb1e917718e4', '_model': 'AutomatedEmail'}, 'cc': [], 'bcc': [], 'body': "Test,\n\nYou've been added to the MAGStock registration database, but we don't have all of your personal information.  To ensure that you can pick up your badge with no hassles at our registration desk, please fill out the rest of your info at https://vagrant-ubuntu-trusty-64.lan:4443/uber/preregistration/confirm?id=157712d4-aa9a-49f3-904e-a59be6933d84 and then simply bring a photo ID to MAGStock.\n\nPlease let us know if you have any questions.\n\nMAGStock Registration Department"}
[2018-05-25 10:27:10,146: INFO/ForkPoolWorker-2] Task uber.tasks.email.send_automated_emails[61e4a966-5b0a-4438-afab-1416fb06a315] succeeded in 0.19316686799720628s: {'badge_confirmation_reminder': 1}
```

**Comment by kitsuta** [comment]
OK, think we've got this figured out.

There's TWO bugs in play: the one I found, which has very similar symptoms to the one reported here, and the one that this report actually describes.

The bug I found (e.g., the one that was affecting my dev instance) was a problem in starting the `uber.tasks.email.send_automated_emails` task. If you restart the celery-beat container (which is responsible for kicking that task off), it gets confused about whether or not it's running and ends up just never doing anything. This means that _no_ automated emails would be sent. There is a fix for this that I'll push out to our quickstart repo.

The bug described here, meanwhile, only affected _some_ automated emails. What was actually happening is that the celery-worker container was running an old, cached copy of the app's code. Celery-worker's version of the server tried to access a DB column that we'd refactored out, and couldn't send any emails after that point because accessing a DB column that doesn't actually exist is a fatal error.

In order to avoid this bug, we need to use the `--build` flag on the celery-worker and web containers when code is updated. That should ensure that the code actually gets updated on both, instead of having some weird caching issue in one of the containers.
